src/interpreter/inner.rs

```diff
@@ -190,6 +201,65 @@ pub fn from_txdata<'txin>(
                 None => Err(Error::UnexpectedStackEnd),
             }
         }
+    // ** pay to taproot **//
+    } else if spk.is_v1_p2tr() {
+        if !ssig_stack.is_empty() {
+            Err(Error::NonEmptyScriptSig)
+        } else {
+            let output_key = bitcoin::XOnlyPublicKey::from_slice(&spk[2..])
+                .map_err(|_| Error::XOnlyPublicKeyParseError)?;
+            if wit_stack.len() == 1 {
+                // Key spend
+                Ok((
+                    Inner::PublicKey(output_key.into(), PubkeyType::Tr),
+                    wit_stack,
+                    None, // Tr script code None
+                ))
+            } else {
+                // wit_stack.len() >=2
+                // Check for annex
+                let ctrl_blk = wit_stack.pop().ok_or(Error::UnexpectedStackEnd)?;
+                let ctrl_blk = ctrl_blk.as_push()?;
+                let tap_script = wit_stack.pop().ok_or(Error::UnexpectedStackEnd)?;
+                if ctrl_blk.len() > 0 && ctrl_blk[0] == TAPROOT_ANNEX_PREFIX {
+                    // Annex is non-standard, bitcoin consensus rules ignore it.
+                    // Our sighash structure and signature verification
+                    // does not support annex, return error
+                    return Err(Error::TapAnnexUnsupported);
+                } else if wit_stack.len() >= 2 {
```

Comment:
this line is incorrect -- wit_stack can be 1 element (or in the case of CTV, empty!)

Reply:
I think the confusion comes from the fact that the tap_script and ctrl_blk are popped off the wit_stack at this point.

Reply:
Yup, the script spend must contain at least two witness elements that are being popped here. The additional witness elements can be 0

Reply:
nope; I still think it's incorrect.

referencing: https://github.com/bitcoin/bitcoin/blob/64a4483dc6798a9a7d8327d320a17b3c4d7d4ee0/src/script/interpreter.cpp#L1919

1. if stack empty, fail
1. Check if there is 2 things and one of the things is an annex, remove it
2. If the stack is length 1 do keypath
3. If the stack is length 2 do script spend

Whereas your code says:

1. if stack empty, fail
1. If the length is 1 do keyspend
2. If the length is >= 2, pop the stack for CTRL and pop for tap script (stack length now 0 to N elemented):
3. If CTRL is actually an annex (0x50), fail.
4. Now control is actually a control, and script is a script for sure
5. check that there are >= 2 elements on the stack; if not fail
6. Execute the script

Clearly, the last step should not be happending! For example, the following witness fails:

<sig> <pk checksig> <ctrl>

because the script only needs one argument.

Reply:
(the right number of arguments is zero)

Reply:
> Clearly, the last step should not be happending! For example, the following witness fails:

Are you missing adding an example?

If I am understanding this correctly, this will fail when the key spend has annex? To be clear, my only concern here is we should fail cleanly on the annex, and not the exact emulation of bitcoin consensus.

Reply:
Added a commit to reflect the first part

Reply:
lol -- example was there, but markdown was hiding it because it looked like html